### DISCUSSION TITLE
Store the last updated date timestamp for each project

### DISCUSSION THREAD

**Discussion by adanielyan** [tolgee/tolgee-platform#1745] [Ideas]
I have a use case when I need to download the translations from tolgee and store them locally in case there is no internet connection. Every time my the app launches I want to refresh the cached translations by downloading them from tolgee, in case they changed since the last caching.  However, my project has a few thousand translation and downloading them takes time, especially for users with slow Internet connection. 

Therefore I want to download translations only if they where changed. For that reason it would be great to have a project "updated" timestamp in the `/v2/projects/{projectId}/stats` API endpoint that will update every time a translation in that project is added, deleted, or updated.

**Comment by JanCizmar** [comment]
Hi there! Thanks for reaching us! That's where ETAG header, comes into play. Just tell us which endpoint do you use for the export and  we can add the support for it.

**Reply by adanielyan** [reply]
Thank you for a super quick response! This is the endpoint we use: `/v2/projects/translations/{languages}`.

**Comment by JanCizmar** [comment]
For `/v2/projects/translations/{languages}` there already is ETag header support implemented. So if you store the ETag and send it in each request, you'll get `Not modified` response, if the data didn't change.

**Reply by adanielyan** [reply]
In which version was it first added? I don't see the ETag header in the API response. Also, I couldn't find any mentions of ETag anywhere in docs. Is it documented anywhere?

**Reply by JanCizmar** [reply]
It is this one: https://github.com/tolgee/tolgee-platform/releases/tag/v2.49.0

**Reply by JanCizmar** [reply]
Oh! Sorry! We are actually using Last Modified header.

**Reply by adanielyan** [reply]
Thank you. I have v3 running in a docker container, but don't get the ETag header. What header should I include in my request to get it?

**Reply by adanielyan** [reply]
Ah, should I send Last-Modified header with my request?

**Reply by JanCizmar** [reply]
Sorry! You should send `If-Modified-Since` header.

**Reply by adanielyan** [reply]
Perfect, thank you so much!

**Reply by adanielyan** [reply]
Sorry, can this header be also added to `/v2/projects/{projectId}` so that we don't have to actually retrieve the translations to know when they where last modified?

**Reply by JanCizmar** [reply]
It cannot, because it's always related to the translations data. Sorry. But the server won't send the data if it wasn't modified after that time. But you can set the last modified header to the time on future, get Not Modified response containing the actual Last Modified value. That should work (I didn't test it, just checked the docs).

**Reply by adanielyan** [reply]
Thank you, Jan. One more question. What format should the value of `If-Modified-Since` header be? Is it a timestamp? I noticed that the `Last-Modified` header is sent in the response even when I don't provide the `If-Modified-Since` header in my request.

**Reply by cyyynthia** [reply]
`Last-Modified` is the information the server is giving you about when the resource was last modified. You should keep this value somewhere, and send it back in the `If-Modified-Since` header. Similar to how `ETag` and `If-None-Match` work.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since

**Reply by adanielyan** [reply]
Thank you, Cynthia. The value of the `Last-Modified` is a formatted date string, e.g. `Fri, 09 Jun 2023 15:49:52 GMT`. Should I send that same formatted string in `If-Modified-Since` or should I send a timestamp, e.g. 1686423028?

**Reply by adanielyan** [reply]
OK, never mind, found my answer in the link you shared. Thanks again!

**Comment by aryanarayantiwari** [comment]
@JanCizmar  does the [Export Data Api](https://docs.tolgee.io/api/export-data) also support ```If-Modified-Since``` header ? If no then what should I use ? if I want tolgee universal placeholders along with ```APPLE_STRINGS_STRINGSDICT``` format .

**Reply by JanCizmar** [reply]
Hello, unfortunately the endpoint doesn't support this. However, you can use the content delivery (CDN), where this is supported.
